allennlp/data/semparse/knowledge_graphs/table_knowledge_graph.py

```diff
@@ -60,14 +57,34 @@ def read_from_json(cls, json_object: Dict[str, Any]) -> 'TableKnowledgeGraph':
         """
         neighbors: DefaultDict[str, List[str]] = defaultdict(list)
         # Following Sempre's convention for naming columns.
-        columns = ["fb:row.row.%s" % cls._normalize_string(x) for x in json_object["columns"]]
+        column_ids = []
+        columns: Dict[str, int] = {}
+        for column_string in json_object['columns']:
+            normalized_string = f'fb:row.row.{cls._normalize_string(column_string)}'
+            if normalized_string in columns:
+                columns[normalized_string] += 1
+                normalized_string = f'{normalized_string}_{columns[normalized_string]}'
+            columns[normalized_string] = 1
+            column_ids.append(normalized_string)
+
+        # TODO(mattg): I know that sempre gives separate ids to columns that collide on
+        # normalization; I don't know if sempre similarly does this for cells that collide.  We'll
+        # implement it anyway.
```

Comment:
If sempre doesn't do it, then you're right, we shouldn't do it.  Ignoring sempre for a minute, though, why should we have cell identifiers that are not unique?

Reply:
Continuing with the same example as above, say we have a question on that table like "How many movies were released in 2012?", then we expect the logical form to be something like `(count (fb:row.row.year fb:cell.2012))`, for which to give the right answer, all the cells that contain the value "2012" should have the same identifier. Why would you want to produce `fb:cell.2012_2`, `fb:cell.2012_3` etc.?

Reply:
Yes, that makes a lot of sense.  But also consider the table:

| Movie | Year |
| --- | --- |
| 1 | 2012 |
| 2 | 2012 |
| 3 | 2012 |

What are the neighbors of the "year" cell?  How many are there?  With our current representation it sure looks like there is one neighbor.  I think there's value in being able to both refer to a cell by some unique identifier, and query for all cells with a given text.

My issue was that I was thinking of `fb:cell.x` as the identifier for a particular cell (as that's pretty much implied by how our knowledge graph code works).  But that's not how sempre defines `fb:cell.x` - it's a function that returns all cells with text `x`.  That's fine, and such a function is useful, it's just odd to be defining table entities that way.

Reply:
In fact, it makes me want to change the type of `fb:cell.x`...  But we can worry about that later.

Reply:
Ah, yes. I see your point and I agree that table entities should be defined differently. We should separate the knowledge graph representation from logical form representation to avoid confusion.

Reply:
I think for now we should just do what sempre does; we can think about doing something smarter when we write a code API for this, and ditch sempre / DPD entirely.